File: coldatomlab/scan3d.py
"""Independent compact scan verification: evolved projections, camera frames, circular statistics."""

import math
from dataclasses import asdict

import numpy as np

from .camera3d import acquire_projection, validate_camera, verify_image
from .interferometry3d import stages
from .solver3d import Config3D, Solver3D


def wrap(value):
    return math.atan2(math.sin(value), math.cos(value))


def statistics(images):
    fits = [v["measured"] for v in images if v["measured"]["available"]]
    count = len(fits)
    result = dict(
        attempted=len(images),
        valid=count,
        failed=len(images) - count,
        failure_fraction=(len(images) - count) / len(images) if images else None,
        mean_phase=None,
        phase_sd=None,
        resultant=None,
        mean_period=None,
        mean_contrast=None,
    )
    if not count:
        return result
    c = sum(math.cos(v["phase"]) for v in fits) / count
    s = sum(math.sin(v["phase"]) for v in fits) / count
    r = min(1, math.hypot(c, s))
    result["resultant"] = r
    if r >= 0.1:
        result["mean_phase"] = math.atan2(s, c)
        if count >= 2:
            result["phase_sd"] = math.sqrt(max(0, -2 * math.log(r)))
    result["mean_period"] = sum(v["spacing"] for v in fits) / count
    result["mean_contrast"] = sum(v["contrast"] for v in fits) / count
    return result


def guide(mode, config):
    return wrap(
        config["relative_phase"]
        if mode == "phase"
        else -config["hold_bias"]
        * math.floor(config["hold_time"] / config["dt"] + 0.5)
        * config["dt"]
    )


def protocol(c):
    split = math.floor(c.split_time / c.dt + 0.5) if c.experiment == "sequence" else 0
    hold = math.floor(c.hold_time / c.dt + 0.5) if c.experiment == "sequence" else 0
    expansion = math.floor(c.duration / c.dt + 0.5)
    ms = c.scales["time_ms"]
    return dict(
        split_steps=split,
        hold_steps=hold,
        expansion_steps=expansion,
        hold_ms=hold * c.dt * ms,
        end_ms=(split + hold + expansion) * c.dt * ms,
    )


def summary(record, mode):
    values = statistics(record["shots"])
    truth = record["ideal"]["measured"]
    g = guide(mode, record["source"]["config"])
    return values | dict(
        guide=g,
        ideal_phase=truth["phase"],
        ideal_period=truth["spacing"],
        ideal_contrast=truth["contrast"],
        camera_minus_ideal=None
        if values["mean_phase"] is None or truth["phase"] is None
        else wrap(values["mean_phase"] - truth["phase"]),
        ideal_minus_guide=None if truth["phase"] is None else wrap(truth["phase"] - g),
    )


def refinement(records):
    result = []
    for coarse in records:
        if coarse["grid"] != 64 or coarse["status"] != "complete":
            continue
        fine = next(
            (
                r
                for r in records
                if r["point"] == coarse["point"] and r["grid"] == 128 and r["status"] == "complete"
            ),
            None,
        )
        if fine is None:
            continue
        a, b = coarse["source"]["diagnostics"], fine["source"]["diagnostics"]
        pa, pb = coarse["ideal"]["measured"], fine["ideal"]["measured"]
        result.append(
            dict(
                point=coarse["point"],
                value=coarse["value"],
                width_relative=[y / x - 1 for x, y in zip(a["widths"], b["widths"])],
                norm_difference=b["norm"] - a["norm"],
                ideal_phase_difference=wrap(pb["phase"] - pa["phase"])
                if pa["available"] and pb["available"]
                else None,
            )
        )
    return result


def compare(saved, expected, path="value"):
    """Reject missing/nonfinite data, including forged null statistics."""
    if isinstance(expected, dict):
        if not isinstance(saved, dict) or saved.keys() != expected.keys():
            raise ValueError(f"Scan {path} keys differ.")
        for key, value in expected.items():
            compare(saved[key], value, f"{path}.{key}")
    elif isinstance(expected, list):
        if not isinstance(saved, list) or len(saved) != len(expected):
            raise ValueError(f"Scan {path} size differs.")
        for a, b in zip(saved, expected):
            compare(a, b, path)
    elif expected is None or isinstance(expected, (str, bool)):
        if saved != expected:
            raise ValueError(f"Scan {path} differs.")
    elif (
        not isinstance(saved, (int, float))
        or isinstance(saved, bool)
        or not math.isfinite(saved)
        or not math.isclose(saved, expected, rel_tol=2e-10, abs_tol=1e-7)
    ):
        raise ValueError(f"Scan {path} differs.")


def jobs(plan):
    mode = plan["mode"]
    if mode not in ("phase", "hold", "bias"):
        raise ValueError("Unknown scan protocol.")
    points, repeats, refine = plan["points"], plan["repeats"], plan["refine"]
    if (
        type(points) is not int
        or not 2 <= points <= 9
        or type(repeats) is not int
        or not 1 <= repeats <= 20
        or type(refine) is not bool
        or points * repeats * (2 if refine else 1) > 80
    ):
        raise ValueError("Invalid scan workload.")
    lo, hi = {"phase": (-math.pi, math.pi), "hold": (0, 3), "bias": (-5, 5)}[mode]
    if (
        not all(type(plan[k]) in (float, int) and math.isfinite(plan[k]) for k in ("start", "end"))
        or not lo <= plan["start"] < plan["end"] <= hi
    ):
        raise ValueError("Invalid scan range.")
    base = Config3D(**plan["base"])
    if (
        base.n != 64
        or base.experiment != ("pair" if mode == "phase" else "sequence")
        or asdict(base) != plan["base"]
    ):
        raise ValueError("Invalid scan base recipe.")
    key = {"phase": "relative_phase", "hold": "hold_time", "bias": "hold_bias"}[mode]
    result = []
    for point in range(points):
        value = plan["start"] + (plan["end"] - plan["start"]) * point / (points - 1)
        for n in [64, 128] if refine else [64]:
            c = Config3D(**(plan["base"] | {key: value, "n": n}))
            validate_camera(plan["camera"], n)
            result.append(dict(point=point, value=value, grid=n, config=asdict(c)))
    return result


def verify_scan(data):
    if (
        data.get("schema") != "coldatomlab-scan3d-v1"
        or data.get("camera_model") != "rb87-browser-camera-v1"
    ):
        raise ValueError("Unknown scan schema/camera model.")
    plan = data["plan"]
    expected_jobs = jobs(plan)
    records = data["records"]
    if (
        data["status"] not in ("complete", "cancelled", "failed")
        or len(records) > len(expected_jobs)
        or (data["status"] == "complete" and len(records) != len(expected_jobs))
    ):
        raise ValueError("Scan completion status/count differs.")
    evidence = []
    for record, job in zip(records, expected_jobs):
        compare(
            {k: record[k] for k in ("point", "value", "grid")},
            {k: job[k] for k in ("point", "value", "grid")},
            "job",
        )
        if record["status"] == "failed":
            if not isinstance(record.get("reason"), str) or not record["reason"]:
                raise ValueError("Stopped scan point needs a reason.")
            # A stopped run has no saved projection: its failure claim cannot be verified.
            evidence.append(dict(point=job["point"], grid=job["grid"], status="unverified_stop"))
            continue
        if record["status"] != "complete":
            raise ValueError("Invalid scan point status.")
        source = record["source"]
        compare(source["config"], job["config"], "source.config")
        c = Config3D(**job["config"])
        end = math.floor(c.duration / c.dt + 0.5) if c.experiment == "pair" else stages(c)[2]
        release = 0 if c.experiment == "pair" else stages(c)[1]
        if source["steps"] != end or source["release_step"] != release:
            raise ValueError("Scan source did not complete the declared protocol.")
        if (
            source["backend"].get("type") != "webgpu"
            or source["backend"].get("precision") != "complex-f32"
        ):
            raise ValueError("Scan source must declare WebGPU float32.")
        compare(source["scales"], c.scales, "source.scales")
        compare(source["protocol"], protocol(c), "source.protocol")
        sim = Solver3D(c)
        while not sim.complete and not sim.warning:
            sim.advance(20)
        if sim.steps != end or sim.warning:
            raise ValueError("CPU reference stopped before the scan endpoint.")
        compare(source["x"], sim.x.tolist(), "source.x")
        snapshot = sim.snapshot()
        columns = np.asarray(source["columns"], dtype=float)
        if columns.shape != (3, c.n, c.n) or not np.isfinite(columns).all() or (columns < 0).any():
            raise ValueError("Invalid scan projections.")
        oracle = np.asarray(snapshot["columns"])
        differences = [
            float(np.linalg.norm(a - b) / np.linalg.norm(b)) for a, b in zip(columns, oracle)
        ]
        diag = source["diagnostics"]
        dx = c.length / c.n
        norm = float(columns[0].sum() * dx**2)
        if not np.allclose(columns.sum(axis=(1, 2)) * dx**2, norm, rtol=1e-10, atol=1e-10):
            raise ValueError("Projection normalizations disagree.")
        # Moments from projections independently certify the saved widths.
        marginals = [columns[0].sum(axis=0), columns[0].sum(axis=1), columns[1].sum(axis=1)]
        widths = []
        for marginal in marginals:
            marginal = marginal / marginal.sum()
            mean = np.dot(marginal, sim.x)
            widths.append(float(np.sqrt(np.dot(marginal, (sim.x - mean) ** 2))))
        compare(
            diag, dict(norm=norm, widths=widths, steps=end, time=end * c.dt), "source.diagnostics"
        )
        width_error = np.asarray(widths) / snapshot["diagnostics"]["widths"] - 1
        if max(differences) > 0.005 or max(abs(width_error)) > 0.005 or abs(norm - 1) > 0.001:
            raise ValueError("Scan source exceeds CPU projection/width/norm tolerances.")
        axis = {"z": (0, ["x", "y"]), "y": (1, ["x", "z"]), "x": (2, ["y", "z"])}[
            plan["camera"]["axis"]
        ]
        density = columns[axis[0]] * c.atoms / c.scales["length_um"] ** 2
        coords = sim.x * c.scales["length_um"]
        images = [record["ideal"], *record["shots"]]
        if len(record["shots"]) != plan["repeats"]:
            raise ValueError("Scan exposure count differs.")
        for i, image in enumerate(images):
            camera = plan["camera"] | (
                dict(noise=False, binning=1, fwhm_um=0)
                if i == 0
                else dict(
                    seed=(plan["camera"]["seed"] + job["point"] * plan["repeats"] + i - 1) % (2**32)
                )
            )
            compare(image["camera"], camera, "camera recipe")
            regenerated = acquire_projection(density, coords, axis[1], camera, c.atoms)
            verify_image(image, regenerated)
        compare(record["summary"], summary(record, plan["mode"]), "summary")
        evidence.append(
            dict(
                point=job["point"],
                grid=c.n,
                status="verified",
                projection_relative_l2=differences,
                width_relative_difference=width_error.tolist(),
                images_verified=len(images),
            )
        )
    compare(data["refinement"], refinement(records), "refinement")
    return dict(
        scan_verified=True,
        scope="Saved complete projections, camera frames and statistics; no full GPU complex field or stopped-run verification.",
        status=data["status"],
        records=evidence,
        tolerances=dict(projection_relative_l2=0.005, width_relative=0.005, norm_drift=0.001),
    )

File: coldatomlab/web/scan3d.js
"use strict";
// Scan orchestration only. Evolution and acquisition remain in their existing solvers.
window.Scan3D = {
  wrap: (v) => Math.atan2(Math.sin(v), Math.cos(v)),
  base(mode) {
    return {
      ...Interferometry3D.defaults,
      n: 64,
      length: mode === "phase" ? 32 : 24,
      dt: mode === "phase" ? 0.004 : 0.006,
      atoms: 2000,
      scattering_nm: mode === "phase" ? 0 : 5.3,
      reference_hz: 30,
      fx_hz: 30,
      fy_hz: 42,
      fz_hz: 21,
      duration: mode === "phase" ? 3 : 2,
      preparation_dt: 0.002,
      experiment: mode === "phase" ? "pair" : "sequence",
    };
  },
  validate(plan) {
    if (!["phase", "hold", "bias"].includes(plan.mode))
      throw Error("Choose a scan protocol.");
    if (!Number.isInteger(plan.points) || plan.points < 2 || plan.points > 9)
      throw Error("Use 2–9 scan points.");
    if (
      !Number.isInteger(plan.repeats) ||
      plan.repeats < 1 ||
      plan.repeats > 20
    )
      throw Error("Use 1–20 exposures per point.");
    if (typeof plan.refine !== "boolean")
      throw Error("Invalid grid comparison setting.");
    if (plan.points * plan.repeats * (plan.refine ? 2 : 1) > 80)
      throw Error(
        "Limit this batch to 80 exposures; reduce points or repeats.",
      );
    const limits = { phase: [-Math.PI, Math.PI], hold: [0, 3], bias: [-5, 5] }[
      plan.mode
    ];
    if (
      ![plan.start, plan.end].every(Number.isFinite) ||
      plan.start >= plan.end ||
      plan.start < limits[0] ||
      plan.end > limits[1]
    )
      throw Error(`Scan range must increase within ${limits.join(" to ")}.`);
    if (
      plan.base.n !== 64 ||
      plan.base.experiment !== (plan.mode === "phase" ? "pair" : "sequence")
    )
      throw Error("Use a 64³ baseline with the matching scan experiment.");
    GPUCloud3D.validate(plan.base);
    for (const job of this.jobs(plan)) {
      GPUCloud3D.validate(job.config);
      AbsorptionCamera3D.validate(plan.camera, job.config.n);
    }
    return plan;
  },
  jobs(plan) {
    const key = {
        phase: "relative_phase",
        hold: "hold_time",
        bias: "hold_bias",
      }[plan.mode],
      out = [];
    for (let point = 0; point < plan.points; point++) {
      const value =
        plan.start + ((plan.end - plan.start) * point) / (plan.points - 1);
      for (const n of plan.refine ? [64, 128] : [64])
        out.push({
          point,
          value,
          grid: n,
          config: { ...plan.base, [key]: value, n },
        });
    }
    return out;
  },
  statistics(images) {
    const fits = images.map((i) => i.measured).filter((f) => f.available),
      count = fits.length;
    const result = {
      attempted: images.length,
      valid: count,
      failed: images.length - count,
      failure_fraction: images.length
        ? (images.length - count) / images.length
        : null,
      mean_phase: null,
      phase_sd: null,
      resultant: null,
      mean_period: null,
      mean_contrast: null,
    };
    if (!count) return result;
    const c = fits.reduce((s, f) => s + Math.cos(f.phase), 0) / count,
      s = fits.reduce((s, f) => s + Math.sin(f.phase), 0) / count,
      R = Math.min(1, Math.hypot(c, s));
    result.resultant = R;
    if (R >= 0.1) result.mean_phase = Math.atan2(s, c);
    if (count >= 2 && R >= 0.1)
      result.phase_sd = Math.sqrt(Math.max(0, -2 * Math.log(R)));
    result.mean_period = fits.reduce((s, f) => s + f.spacing, 0) / count;
    result.mean_contrast = fits.reduce((s, f) => s + f.contrast, 0) / count;
    return result;
  },
  guide(mode, c) {
    return this.wrap(
      mode === "phase"
        ? c.relative_phase
        : -c.hold_bias * Math.round(c.hold_time / c.dt) * c.dt,
    );
  },
  protocol(c) {
    const split =
        c.experiment === "sequence" ? Math.round(c.split_time / c.dt) : 0,
      hold = c.experiment === "sequence" ? Math.round(c.hold_time / c.dt) : 0,
      expansion = Math.round(c.duration / c.dt),
      ms = 1000 / (2 * Math.PI * c.reference_hz);
    return {
      split_steps: split,
      hold_steps: hold,
      expansion_steps: expansion,
      hold_ms: hold * c.dt * ms,
      end_ms: (split + hold + expansion) * c.dt * ms,
    };
  },
  summary(record, mode) {
    if (record.status !== "complete") return null;
    const stats = this.statistics(record.shots),
      truth = record.ideal.measured;
    return {
      ...stats,
      guide: this.guide(mode, record.source.config),
      ideal_phase: truth.phase,
      ideal_period: truth.spacing,
      ideal_contrast: truth.contrast,
      camera_minus_ideal:
        stats.mean_phase === null || truth.phase === null
          ? null
          : this.wrap(stats.mean_phase - truth.phase),
      ideal_minus_guide:
        truth.phase === null
          ? null
          : this.wrap(truth.phase - this.guide(mode, record.source.config)),
    };
  },
  refinement(records) {
    return records
      .filter((r) => r.grid === 64 && r.status === "complete")
      .flatMap((r) => {
        const fine = records.find(
          (f) =>
            f.point === r.point && f.grid === 128 && f.status === "complete",
        );
        if (!fine) return [];
        const a = r.source.diagnostics,
          b = fine.source.diagnostics;
        return [
          {
            point: r.point,
            value: r.value,
            width_relative: b.widths.map((v, i) => v / a.widths[i] - 1),
            norm_difference: b.norm - a.norm,
            ideal_phase_difference:
              r.ideal.measured.available && fine.ideal.measured.available
                ? this.wrap(fine.ideal.measured.phase - r.ideal.measured.phase)
                : null,
          },
        ];
      });
  },
  csv(report) {
    const rows = [
      [
        "parameter",
        "parameter_unit",
        "point",
        "value",
        "grid",
        "reference_hz",
        "actual_hold_ms",
        "actual_end_ms",
        "status",
        "valid_exposures",
        "failed_exposures",
        "failure_fraction",
        "mean_phase_rad",
        "circular_sd_rad",
        "ideal_phase_rad",
        "guide_rad",
        "camera_minus_ideal_rad",
        "ideal_minus_guide_rad",
        "mean_period_um",
        "mean_contrast",
        "reason",
      ],
    ];
    for (const r of report.records) {
      const s = r.summary || {};
      rows.push([
        report.plan.mode,
        { phase: "rad", hold: "omega0^-1", bias: "hbar omega0" }[
          report.plan.mode
        ],
        r.point,
        r.value,
        r.grid,
        report.plan.base.reference_hz,
        r.source?.protocol.hold_ms,
        r.source?.protocol.end_ms,
        r.status,
        s.valid,
        s.failed,
        s.failure_fraction,
        s.mean_phase,
        s.phase_sd,
        s.ideal_phase,
        s.guide,
        s.camera_minus_ideal,
        s.ideal_minus_guide,
        s.mean_period,
        s.mean_contrast,
        r.reason || "",
      ]);
    }
    return rows
      .map((r) =>
        r.map((v) => `"${String(v ?? "").replaceAll('"', '""')}"`).join(","),
      )
      .join("\r\n");
  },
};

window.Scan3DRunner = class {
  constructor(
    plan,
    onProgress = () => {},
    factory = (c, p) => GPUCloud3D.create(c, p),
  ) {
    this.plan = structuredClone(Scan3D.validate(plan));
    this.onProgress = onProgress;
    this.factory = factory;
    this.paused = false;
    this.cancelled = false;
    this.report = {
      schema: "coldatomlab-scan3d-v1",
      version: "0.14.0",
      camera_model: "rb87-browser-camera-v1",
      plan: this.plan,
      status: "running",
      records: [],
      refinement: [],
    };
  }
  emit(message) {
    this.onProgress(message, this.report);
  }
  async checkpoint() {
    if (this.paused && !this.cancelled)
      this.emit("Paused at a checkpoint; Resume or Cancel.");
    while (this.paused && !this.cancelled)
      await new Promise((r) => setTimeout(r, 50));
    if (this.cancelled)
      throw new DOMException(
        "Scan cancelled; completed points are retained.",
        "AbortError",
      );
  }
  async run() {
    try {
      const jobs = Scan3D.jobs(this.plan);
      for (let index = 0; index < jobs.length; index++) {
        await this.checkpoint();
        const job = jobs[index];
        let sim = null;
        const title = `Point ${job.point + 1}/${this.plan.points} · ${job.grid}³`;
        this.emit(`${title} · preparing`);
        try {
          sim = await this.factory(job.config, (m) =>
            this.emit(`${title} · ${m}`),
          );
          while (!sim.complete && !sim.warning) {
            await this.checkpoint();
            await sim.advance(20);
            this.emit(`${title} · evolving · step ${sim.steps}`);
            await new Promise((r) => setTimeout(r, 0));
          }
          await this.checkpoint();
          if (sim.warning) {
            this.report.records.push({
              point: job.point,
              value: job.value,
              grid: job.grid,
              status: "failed",
              reason: sim.warning,
            });
            this.emit(`${title} · stopped: ${sim.warning}`);
            continue;
          }
          const snap = sim.snapshot(),
            source = {
              config: structuredClone(snap.config),
              scales: snap.scales,
              x: snap.x,
              columns: snap.columns,
              diagnostics: {
                norm: snap.diagnostics.norm,
                widths: snap.diagnostics.widths,
                time: snap.diagnostics.time,
                steps: snap.diagnostics.steps,
              },
              steps: sim.steps,
              release_step: sim.releaseStep,
              protocol: Scan3D.protocol(snap.config),
              backend: snap.backend,
            };
          const ideal = AbsorptionCamera3D.acquire(snap, {
            ...this.plan.camera,
            binning: 1,
            fwhm_um: 0,
            noise: false,
          });
          const record = {
            point: job.point,
            value: job.value,
            grid: job.grid,
            status: "complete",
            source,
            ideal,
            shots: [],
          };
          for (let repeat = 0; repeat < this.plan.repeats; repeat++) {
            await this.checkpoint();
            this.emit(`${title} · exposure ${repeat + 1}/${this.plan.repeats}`);
            await new Promise((r) => setTimeout(r, 0));
            const seed =
              (this.plan.camera.seed +
                job.point * this.plan.repeats +
                repeat) >>>
              0;
            record.shots.push(
              AbsorptionCamera3D.acquire(snap, { ...this.plan.camera, seed }),
            );
          }
          record.summary = Scan3D.summary(record, this.plan.mode);
          this.report.records.push(record);
          this.report.refinement = Scan3D.refinement(this.report.records);
          this.emit(`${title} · recorded`);
        } finally {
          sim?.destroy();
        }
      }
      this.report.status = "complete";
    } catch (e) {
      this.report.status = e.name === "AbortError" ? "cancelled" : "failed";
      this.report.message = e.message;
    }
    this.emit(
      this.report.status === "complete"
        ? "Scan complete."
        : this.report.message,
    );
    return this.report;
  }
};
